autoscale-groups-core/src/main/java/fish/payara/extensions/autoscale/groups/core/admin/ScaleCommandHelper.java

```diff
@@ -0,0 +1,327 @@
+        ClusterCommandHelper clusterCommandHelper = new ClusterCommandHelper(domain, commandRunner);
+        targetServers = clusterCommandHelper.optimizeServerListOrder(targetServers);
+
+        // Holds responses from the threads running the command
+        ArrayBlockingQueue<CommandRunnable> responseQueue = new ArrayBlockingQueue<>(nInstances);
+
+        // Make the thread pool use the smaller of the number of instances or half the admin thread pool size
+        int threadPoolSize = Math.min(nInstances, getAdminThreadPoolSize() / 2);
+        if (threadPoolSize < 1) {
+            threadPoolSize = 1;
+        }
+
+        ExecutorService threadPool = Executors.newFixedThreadPool(threadPoolSize);
+
+        if (parameterMap == null) {
+            parameterMap = new ParameterMap();
+        }
+
+        LOGGER.info(String.format(
+                "Executing %s on %d instances using a thread pool of size %d: %s", commandName, nInstances,
+                threadPoolSize, targetServers.stream().map(Server::getName).collect(Collectors.joining(", "))));
+
+        progressStatus.setTotalStepCount(nInstances);
+        progressStatus.progress(Strings.get("cluster.command.executing", commandName, nInstances));
+
+        // Loop through instance names, construct the command for each instance name, and hand it off to the threadpool.
+        for (Server server : targetServers) {
+            String serverName = server.getName();
+            waitingForServerNames.add(serverName);
+
+            ParameterMap instanceParameterMap = new ParameterMap(parameterMap);
+            // Set the instance name as the operand for the commnd
+            instanceParameterMap.set("DEFAULT", serverName);
+
+            ActionReport instanceReport = actionReport.addSubActionsReport();
+            instanceReport.setActionExitCode(SUCCESS);
+            CommandRunner.CommandInvocation invocation = commandRunner.getCommandInvocation(commandName, instanceReport,
+                    subject);
+            invocation.parameters(instanceParameterMap);
+
+            String msg = commandName + " " + serverName;
+            LOGGER.info(msg);
+
+            // Wrap the command invocation in a runnable and hand it off
+            // to the thread pool
+            CommandRunnable cmdRunnable = new CommandRunnable(invocation, instanceReport, responseQueue);
+            cmdRunnable.setName(serverName);
+            threadPool.execute(cmdRunnable);
+        }
+
+        if (LOGGER.isLoggable(FINE)) {
+            LOGGER.fine(String.format("%s commands queued, waiting for responses", commandName));
+        }
+
+        // Make sure we don't wait longer than the admin read timeout. Set our limit to be 3 seconds less.
+        long adminTimeout = RemoteRestAdminCommand.getReadTimeout() - 3000;
+        if (adminTimeout <= 0) {
+            // This should never be the case
+            adminTimeout = 57 * 1000;
+        }
```

Comment:
What is the purpose of this validation? To prevent the admin read time timeout from timing out the complete operation? Also, I don't trust that there's a statement there that "should never be the case" that resets the timeout value to a full minute

Reply:
It was taken from [ClusterCommadHelper](https://github.com/payara/Payara/blob/master/nucleus/cluster/admin/src/main/java/com/sun/enterprise/v3/admin/cluster/ClusterCommandHelper.java).

And yes, I believe that's the intention, since this may take longer than a usual timeout depending on how many commands are being run (threadpool size isn't infinite). I guess the fallback timeout of being just less than a minute is a catch all, since it implies your admin timeout is 3 seconds or less (pretty damn short). I suppose you could try and be clever and do something like "80% of admin timeout", but it seems a minor point

Reply:
Fair enough, but it would be good to revisit this code in the future to improve the timeout management.